src/coreclr/jit/assertionprop.cpp

```diff
@@ -4863,9 +4863,28 @@ bool Compiler::optNonNullAssertionProp_Ind(ASSERT_VALARG_TP assertions, GenTree*
 {
     assert(indir->OperIsIndir());
 
-    if (!(indir->gtFlags & GTF_EXCEPT))
+    bool updated = false;
+    if (indir->OperIs(GT_STOREIND))
     {
-        return false;
+        // We like CSE to happen for handles, as the codegen for loading a 64-bit constant can be pretty heavy
+        // and this is particularly true on platforms with a fixed-width instruction encoding. However, this
+        // pessimizes stores as we can no longer optimize around some object handles that would allow us to
+        // bypass the write barrier.
+        //
+        // In order to handle that, we'll propagate the IND_TGT_NOT_HEAP flag onto the store if the handle is
+        // directly or if the underlying value number is an applicable object handle.
+
+        const GenTree* data = indir->AsIndir()->Data();
+        if (data->IsIconHandle(GTF_ICON_OBJ_HDL) || vnStore->IsVNObjHandle(data->gtVNPair.GetConservative()))
+        {
+            indir->gtFlags |= GTF_IND_TGT_NOT_HEAP;
+            updated = true;
+        }
+    }
```

Comment:
This ended up being a simple workaround to the issue described above using the existing flag.

It even fixes some cases that we weren't handling today and ensures that we can continue using the CSE, without relying on LSRA making propagated constants reused itself.

Reply:
(Courtesy EgorBo, from a discussion on Discord)